File: SudokuSolver.js
let puzzle = [
    [5,3,0,0,7,0,0,0,0],
    [6,0,0,1,9,5,0,0,0],
    [0,9,8,0,0,0,0,6,0],
    [8,0,0,0,6,0,0,0,3],
    [4,0,0,8,0,3,0,0,1],
    [7,0,0,0,2,0,0,0,6],
    [0,6,0,0,0,0,2,8,0],
    [0,0,0,4,1,9,0,0,5],
    [0,0,0,0,8,0,0,7,9]];

sudoku(puzzle);

function sudoku(puzzle) {

    let numbersCount = [0, 0, 0, 0, 0, 0, 0, 0, 0];
    let maxNumbers = new Array();
    let empty = 81;
    let found = false;

    // Get count of all numbers
    for (let row = 0; row < 9; row++) {
        for (let col = 0; col < 9; col++) {
            if (puzzle[row][col] != 0) {
                numbersCount[puzzle[row][col] - 1]++;
                empty--;
            }
        }
    }

    while (empty > 0) {
        let boxPuzzle = new Array();
        for (let puz = 0; puz < 9; puz++) {
            boxPuzzle.push(puzzle[puz].slice());
        }

        // Check which number is most present and not blocking
        let max = 0;
        let maxNumber = 0;
        for (let i = 0; i < 9; i++) {
            if (numbersCount[i] > max && !maxNumbers.includes(i + 1) && numbersCount[i] < 9) {
                max = numbersCount[i];
                maxNumber = i + 1;
            }
        }
        let isInRow = false;
        let numberCheck = false;
        let colCheck = 0;
        let rowCheck = 0;

        // Fill non valid spaces with X in the puzzle clone
        for (let row = 0; row < 9; row++) {
            isInRow = false;
            numberCheck === false;
            
            for (let col = 0; col < 9; col++) {

                if (!isInRow && puzzle[row][col] === maxNumber) {
                    isInRow = true;
                    for (let p = 0; p < 9; p++) {
                        if (p != col) {
                            boxPuzzle[row][p] = "X";
                        }
                        if (p != row) {
                            boxPuzzle[p][col] = "X";
                        }
                    }
                    break;
                }
            }
        }

        // Check each mini-puzzle
        for (let box = 0; box < 9; box++) {
            numberCheck === false;
            let nextBox = false;
            let rowStart = 3 * Math.trunc(box / 3);
            let colStart = (box - rowStart) * 3;
            
            for (let row = rowStart; row < rowStart + 3; row++) {

                if (nextBox) {break;};

                for (let col = colStart; col < colStart + 3; col++) {

                    if (boxPuzzle[row][col] === 0) {
                        if (!numberCheck) {
                            numberCheck = true;
                            rowCheck = row;
                            colCheck = col;
                        }
                        else {
                            numberCheck = false;
                            nextBox = true;
                            break;
                        }
                    }
                    if (puzzle[row][col] === maxNumber) {
                        numberCheck = false;
                        nextBox = true;
                        break;
                    }
                }
            }

            if (numberCheck) {
                puzzle[rowCheck][colCheck] = maxNumber;
                numbersCount[maxNumber - 1]++;
                empty--;
                found = true;
            }
        }
       
        if (!found) {
           maxNumbers.push(maxNumber);
        }
        else {
            maxNumbers = [];
            found = false;
        }
    }
    return puzzle;
}

File: sudokuSolver.py
import math
import copy

def sudoku(puzzle) :

    numbersCount = [0, 0, 0, 0, 0, 0, 0, 0, 0]
    maxNumbers = []
    empty = 81
    found = False

    # Get count of all numbers
    for row  in range(9) :
        for col in range(9) :
            if puzzle[row][col] != 0 :
                numbersCount[puzzle[row][col] - 1] += 1
                empty -= 1

    while empty > 0 :
        boxPuzzle = []
        for puz in range(9) :
            boxPuzzle.append(puzzle[puz][:])

        # Check which number is most present and not blocking
        max = 0
        maxNumber = 0
        for i in range(9) :
            if numbersCount[i] > max and (i + 1) not in maxNumbers and numbersCount[i] < 9 :
                max = numbersCount[i]
                maxNumber = i + 1
                
        isInRow = False
        numberCheck = False
        colCheck = 0
        rowCheck = 0

        # Fill non valid spaces with X in the puzzle clone
        for row in range (9) :
            isInRow = False;
            
            for col in range(9) :

                if not isInRow and puzzle[row][col] == maxNumber :
                    isInRow = True
                    for p in range(9) :
                        if p != col :
                            boxPuzzle[row][p] = "X"
                        if p != row :
                            boxPuzzle[p][col] = "X"
                    break

        # Check each mini-puzzle
        for box in range(9) :
            nextBox = False
            rowStart = 3 * math.floor(box / 3)
            colStart = (box - rowStart) * 3
            
            for row in range(rowStart, rowStart + 3) :

                if nextBox :
                    break

                for col in range(colStart, colStart + 3) :

                    if boxPuzzle[row][col] == 0 :
                        if not numberCheck :
                            numberCheck = True
                            rowCheck = row
                            colCheck = col
                        else :
                            numberCheck = False
                            nextBox = True
                            break
                    if puzzle[row][col] == maxNumber :
                        numberCheck = False
                        nextBox = True
                        break

            if numberCheck :
                print("Correct")
                puzzle[rowCheck][colCheck] = maxNumber
                numbersCount[maxNumber - 1] += 1
                empty -= 1
                found = True
                
        if not found :
           maxNumbers.append(maxNumber)
        else :
            maxNumbers = []
            found = False
    return puzzle